### PR TITLE
SI-8207 Allow import qualified by self reference

### PR DESCRIPTION
**Pull request by retronym**
This regressed in SI-6815 / #2374. We check if the result of
`typedQualifier(Ident(selfReference))` is a stable identifier
pattern. But we actually see the expansion to `C.this`, which
doesn't qualify.

This commit adds a special cases to `importSig` to compensate.
This is safe enough, because the syntax prevents the following:

```
scala> class C { import C.this.toString }
<console>:1: error: '.' expected but '}' found.
       class C { import C.this.toString }
                                        ^
```

So loosening the check here doesn't admit invalid programs.
I've backed this up with a `neg` test.

The enclosed test also checks that we can use the self
reference in a singleton type. (That wasn't broken.)

Maybe it would be more principled to avoid expanding the self
reference in `typedIdent`. I can imagine that the current situation
is a pain for refactoring tools that try to implement a rename
refactoring, for example.

Seems a bit risky at the minute, but I've noted the idea
in a comment.

Review by @adriaanm. 

/cc @misto for discussion about the expansion of the `Ident` to a `This` in the IDE context.

### CODE REVIEW

File: test/files/neg/t8207.scala [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class C { import C.this.toString }
+
+class D { import D.this.toString }
```
Comment (paulp) [inline test/files/neg/t8207.scala:3]: Maybe this originally tested something different but degenerated into this smallish delta from class C.

Comment (xeno-by) [inline test/files/neg/t8207.scala:3]: Maybe forgotten 'extends C'?

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 2]
```diff
@@ -1407,8 +1407,14 @@ trait Namers extends MethodSynthesis {
       if (expr1.isErrorTyped)
         ErrorType
       else {
-        if (!treeInfo.isStableIdentifierPattern(expr1))
-          typer.TyperErrorGen.UnstableTreeError(expr1)
+        expr1 match {
+          case This(_) =>
+            // SI-8207 okay, typedIdent expands Ident(self) to C.this which doesn't satisfy the next case
+            // TODO should we change `typedIdent` not to expand to the `Ident` to a `This`?
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:1413]: Why does a self-reference expand to this?

**Comment by adriaanm** [comment]
Not sure, it seems like the ident's symbol should be able to tell us it's a this-reference just as easily.

File: test/files/neg/t8207.scala (hunk 1 above)
Comment (retronym) [inline test/files/neg/t8207.scala:3]: I guess I meant `this`. Juggling too many tickets today.

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala (hunk 2 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:1413]: A quick inspection suggests its done by the following in `Context#lookupSymbol`:

``` scala
def finishDefSym(sym: Symbol, pre0: Type): NameLookup =
        if (requiresQualifier(sym))
          finish(gen.mkAttributedQualifier(pre0), sym)
        else
          finish(EmptyTree, sym)

      def requiresQualifier(s: Symbol) = (
           s.owner.isClass
        && !s.owner.isPackageClass
        && !s.isTypeParameterOrSkolem
      )
```

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**